freshmaker/pyxis.py

```diff
@@ -0,0 +1,246 @@
+
+import requests
+import urllib
+import copy
+from requests_kerberos import HTTPKerberosAuth, OPTIONAL
+from packaging import version
+
+from freshmaker import log, conf
+
+
+class PyxisRequestError(Exception):
+    """
+    Error return as a response from Pyxis
+    """
+
+    def __init__(self, status_code, error_response):
+        """
+        Initialize Pyxis request error
+
+        :param int status_code: response status code
+        :param str or dict error_response: response content returned from Pyxis
+        """
+
+        self._status_code = status_code
+        self._raw = error_response
+
+    @property
+    def raw(self):
+        return self._raw
+
+    @property
+    def status_code(self):
+        return self._status_code
+
+
+class Pyxis(object):
+    """ Interface for querying Pyxis"""
+
+    def __init__(self, server_url):
+        self._server_url = server_url
+        # add api version to root url
+        self._api_root = urllib.parse.urljoin(self._server_url, "v1/")
+
+    def _make_request(self, entity, params):
+        """
+        Send a request to Pyxis
+
+        :param str entity: entity part to construct a full URL for request.
+        :param dict params: Pyxis query parameters.
+        :return: Json response from Pyxis
+        :rtype: dict
+        :raises PyxisRequestError: If Pyxis returns error as a response
+        """
+        entity_url = urllib.parse.urljoin(self._api_root, entity)
+
+        auth_method = HTTPKerberosAuth(mutual_authentication=OPTIONAL)
+        response = requests.get(entity_url, params=params, auth=auth_method,
+                                timeout=conf.net_timeout)
+
+        if response.ok:
+            return response.json()
+
+        # Warn early, in case there is an error in the error handling code below
+        log.warning("Request to %s gave %r", response.request.url, response)
+
+        try:
+            response_text = response.json()
+        except ValueError:
+            response_text = response.text
+
+        raise PyxisRequestError(response.status_code, response_text)
+
+    def _pagination(self, entity, params):
+        """
+        Process all pages in Pyxis
+
+        :param str entity: what data/entity to request from Pyxis
+        :param dict params: parameters to add to GET request
+        :return: list of all 'data' fields from responses from Pyxis
+        :rtype: list
+        """
+        local_params = {"page_size": "100"}
+        local_params.update(params)
+        ret = []
+        page = 0
+        while True:
+            local_params["page"] = page
+            response_data = self._make_request(entity, params=local_params)
+            # When the page after the actual last page is reached, data will be an empty list
+            if not response_data.get('data'):
+                break
+            ret.extend(response_data['data'])
+            page += 1
+
+        return ret
+
+    def get_operator_indices(self, params=None):
```

Comment:
It seems `params` isn't ever used. Could you please remove it? It's best to not add unused code. We can always add it when needed.